### ISSUE TITLE
Check failed randomly on aircrack-ng Harkonen

### ISSUE THREAD

**Issue by aircrack-ng** [aircrack-ng/aircrack-ng#1680]
**Reported by sophie on 19 Feb 2016 10:26 UTC**

we packaged the version 1.2-RC4 for kali linux and the build fails sometimes on this check:
```
./aircrack-ng -w ../test/password.lst -a 2 -e Harkonen -q ../test/wpa2.eapol.cap | grep 'KEY FOUND! \[ 12345678 \]' 
```
with the following error: "Passphrase not in dictionary"

We had failures on i386 and amd64. We need to build the package several times until it succeeds.

**Comment by aircrack-ng** [comment]
**Comment by sophie on 19 Feb 2016 10:26 UTC**

We had this problem at least on amd64 with common.cfg like this:
```
AVX2FLAG=Y
SIMDCORE=true
INTEL_ASM=Y
PTHREAD=Y
```

Note however that we cannot reproduce the problem with this common.cfg on i386:
```
NEWSSE=false
SIMDCORE=false
PTHREAD=Y
```

On amd64 with that same common.cfg I get a segfault:
```
(gdb) run -w ../test/password.lst -a 2 -e Harkonen -q ../test/wpa2.eapol.cap
Starting program: /home/sophie/kali/Packaging/aircrack-ng/src/aircrack-ng -w ../test/password.lst -a 2 -e Harkonen -q ../test/wpa2.eapol.cap
[debugging using libthread_db enabled](Thread)
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[Thread 0x7ffff65b4700 (LWP 7718)](New)
[0x7ffff65b4700 (LWP 7718) exited](Thread)
[Thread 0x7ffff65b4700 (LWP 7719)](New)
[Thread 0x7ffff5db3700 (LWP 7720)](New)
[Thread 0x7ffff55b2700 (LWP 7721)](New)
[Thread 0x7ffff4db1700 (LWP 7722)](New)
[Thread 0x7fffeffff700 (LWP 7723)](New)

Program received signal SIGSEGV, Segmentation fault.
[to Thread 0x7ffff5db3700 (LWP 7720)](Switching)
shasse2_ends () at sha1-sse2.S:340
340	        movdqa   MANGLE(const_ff00), tmp3
(gdb) bt
#0  shasse2_ends () at sha1-sse2.S:340
#1  0x000000004cadc1ef in ?? ()
#2  0x0000555555559f6f in calc_4pmk (_key1=<optimized out>, _key2=<optimized out>, 
    _key3=<optimized out>, _key4=<optimized out>, _essid=<optimized out>, _pmk1=0x7ffff5db2a40 "", 
    _pmk2=0x7ffff5db2ac0 "", _pmk3=0x7ffff5db2b40 "", _pmk4=0x7ffff5db2bc0 "") at sha1-sse2.h:193
#3  0x0000555555562397 in crack_wpa_thread (arg=<optimized out>) at aircrack-ng.c:4152
#4  0x00007ffff7bc5284 in start_thread (arg=0x7ffff5db3700) at pthread_create.c:333
#5  0x00007ffff6dba97d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
```

**Comment by aircrack-ng** [comment]
**Comment by misterx on 19 Feb 2016 10:26 UTC**

What does ```aircrack-ng -u``` reports for the build system?

**Comment by aircrack-ng** [comment]
**Comment by misterx on 19 Feb 2016 10:26 UTC**

I'm not entirely sure to understand because in the ticket you say you have failures in i386 and amd64 and in the update you say it's fine on i386. Could you provide more details?

**Comment by aircrack-ng** [comment]
**Comment by DaKahuna on 19 Feb 2016 10:26 UTC**

 I am unable to duplicate this error. 

 root@kali:~# uname -a
Linux kali 4.0.0-kali1-amd64 #1 SMP Debian 4.0.4-1+kali2 (2015-06-03) x86_64 GNU/Linux
root@kali:~# aircrack-ng -a 2 -e Avalanche -w /usr/share/wordlists/rockyou.txt /media/root/USB30FD/test/home-chan-1-01.cap 

  Aircrack-ng 1.2 rc2


                   [8120 keys tested (4340.21 k/s)


                           KEY FOUND! [ Jasmine1 ](00:00:01])


      Master Key     : E3 5B 3E F7 A3 E7 44 B5 0C 0C C9 EA FA 77 02 84 
                       09 79 C4 9D A4 43 53 06 ED CE C2 20 01 29 C7 CD 

      Transient Key  : CE DC 5D 17 4B 9E F8 BB 27 F8 4D D7 2C 35 DE 73 
                       6D 38 1C BF AE 2E 11 2E EA 4A 5F 74 7A F7 EE E7 
                       08 B2 9A A5 2D 8B 6D E4 9A A4 FB A0 89 67 67 C2 
                       A8 F9 62 D1 3A E0 40 61 61 39 61 62 F3 30 C6 2C 

      EAPOL HMAC     : 46 53 AB B4 7F DC 73 05 DB 49 1E 2F 03 9F 33 0D 
root@kali:~#

**Comment by aircrack-ng** [comment]
**Comment by DaKahuna on 19 Feb 2016 10:26 UTC**


 Also --

 root@kali:~# aircrack-ng -a 2 -e Avalanche -w /usr/share/wordlists/rockyou.txt /media/root/USB30FD/test/home-chan-1-01.cap -q 
KEY FOUND! [ Jasmine1 ]
root@kali:~#

**Comment by aircrack-ng** [comment]
**Comment by DaKahuna on 19 Feb 2016 10:26 UTC**

 Downloaded and installed aircrack-ng-1.2-rc4; 
 Installed libssl-dev;
 compiled and ran successfully.

root@kali:~# aircrack-ng -a 2 -e Avalanche -w /usr/share/wordlists/rockyou.txt /media/root/USB30FD/test/home-chan-1-01.cap 

  Aircrack-ng 1.2 rc3


                   [8104 keys tested (4333.98 k/s)


                           KEY FOUND! [ Jasmine1 ](00:00:01])


      Master Key     : E3 5B 3E F7 A3 E7 44 B5 0C 0C C9 EA FA 77 02 84 
                       09 79 C4 9D A4 43 53 06 ED CE C2 20 01 29 C7 CD 

      Transient Key  : CE DC 5D 17 4B 9E F8 BB 27 F8 4D D7 2C 35 DE 73 
                       6D 38 1C BF AE 2E 11 2E EA 4A 5F 74 7A F7 EE E7 
                       08 B2 9A A5 2D 8B 6D E4 9A A4 FB A0 89 67 67 C2 
                       A8 F9 62 D1 3A E0 40 61 61 39 61 62 F3 30 C6 2C 

      EAPOL HMAC     : 46 53 AB B4 7F DC 73 05 DB 49 1E 2F 03 9F 33 0D

**Comment by aircrack-ng** [comment]
**Comment by misterx on 19 Feb 2016 10:26 UTC**

make check works just fine and the only other place (other than your system) I see it fail in on Travis. Could you provide ```aircrack-ng -u``` as well as the output of ```cat /proc/cpuinfo``` or ```lscpu```

**Comment by aircrack-ng** [comment]
**Comment by sophie on 19 Feb 2016 10:26 UTC**

Hello,

sorry for the delay, here are the requested information. This is within a Kali Rolling chroot. The kernel is Linux 4.3.5 from Debian. The 

```
(kali-dev-amd64-test) x250-sophie:~/kali/Packaging/aircrack-ng
(2003)$ cat common.cfg 
AVX2FLAG=Y
SIMDCORE=true
INTEL_ASM=Y
PTHREAD=Y
(kali-dev-amd64-test) x250-sophie:~/kali/Packaging/aircrack-ng/src
(1999)$ ./aircrack-ng -u
Vendor          = Intel
Model           = Intel(R) Core(TM) i7-5600U CPU @ 2.60GHz
Features        = MMX,SSE,SSE2,SSE3,SSSE3,SSE4.1,SSE4.2,AES-NI,AVX,AVX2
Hyper-Threading = No
CPU frequency   = 800 MHz (Max: 3200 MHz)
CPU cores       = 4
SIMD size       = 8 (256 bit)
(kali-dev-amd64-test) x250-sophie:~/kali/Packaging/aircrack-ng/src
(2000)$ cat /proc/cpuinfo 
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 61
model name	: Intel(R) Core(TM) i7-5600U CPU @ 2.60GHz
stepping	: 4
microcode	: 0x22
cpu MHz		: 2600.812
cache size	: 4096 KB
physical id	: 0
siblings	: 4
core id		: 0
cpu cores	: 2
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 20
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch ida arat epb pln pts dtherm intel_pt tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt
bugs		:
bogomips	: 5187.83
clflush size	: 64
cache_alignment	: 64
address sizes	: 39 bits physical, 48 bits virtual
power management:

processor	: 1
vendor_id	: GenuineIntel
cpu family	: 6
model		: 61
model name	: Intel(R) Core(TM) i7-5600U CPU @ 2.60GHz
stepping	: 4
microcode	: 0x22
cpu MHz		: 2624.679
cache size	: 4096 KB
physical id	: 0
siblings	: 4
core id		: 0
cpu cores	: 2
apicid		: 1
initial apicid	: 1
fpu		: yes
fpu_exception	: yes
cpuid level	: 20
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch ida arat epb pln pts dtherm intel_pt tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt
bugs		:
bogomips	: 5187.83
clflush size	: 64
cache_alignment	: 64
address sizes	: 39 bits physical, 48 bits virtual
power management:

processor	: 2
vendor_id	: GenuineIntel
cpu family	: 6
model		: 61
model name	: Intel(R) Core(TM) i7-5600U CPU @ 2.60GHz
stepping	: 4
microcode	: 0x22
cpu MHz		: 2600.203
cache size	: 4096 KB
physical id	: 0
siblings	: 4
core id		: 1
cpu cores	: 2
apicid		: 2
initial apicid	: 2
fpu		: yes
fpu_exception	: yes
cpuid level	: 20
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch ida arat epb pln pts dtherm intel_pt tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt
bugs		:
bogomips	: 5187.83
clflush size	: 64
cache_alignment	: 64
address sizes	: 39 bits physical, 48 bits virtual
power management:

processor	: 3
vendor_id	: GenuineIntel
cpu family	: 6
model		: 61
model name	: Intel(R) Core(TM) i7-5600U CPU @ 2.60GHz
stepping	: 4
microcode	: 0x22
cpu MHz		: 2751.328
cache size	: 4096 KB
physical id	: 0
siblings	: 4
core id		: 1
cpu cores	: 2
apicid		: 3
initial apicid	: 3
fpu		: yes
fpu_exception	: yes
cpuid level	: 20
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch ida arat epb pln pts dtherm intel_pt tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt
bugs		:
bogomips	: 5187.83
clflush size	: 64
cache_alignment	: 64
address sizes	: 39 bits physical, 48 bits virtual
power management:

(kali-dev-amd64-test) x250-sophie:~/kali/Packaging/aircrack-ng/src
(2001)$ uname -a
Linux x250-sophie 4.3.0-1-amd64 #1 SMP Debian 4.3.5-1 (2016-02-06) x86_64 GNU/Linux

```

**Comment by aircrack-ng** [comment]
**Comment by sophie on 19 Feb 2016 10:26 UTC**

Replying to [misterx](comment:3):
> I'm not entirely sure to understand because in the ticket you say you have failures in i386 and amd64 and in the update you say it's fine on i386. Could you provide more details?

When we left the build system generate the common.cfg file we had failures on amd64 and i386 like we described. Then we tried to find a work-around by tweaking common.cfg manually and then we discovered what we explained in our follow up message... we found a working config for i386 but which was completely not working for amd64 (segfault).

**Comment by aircrack-ng** [comment]
**Comment by misterx on 19 Feb 2016 10:26 UTC**

Which distribution do you use to build packages? 32 or 64 bit?

**Comment by aircrack-ng** [comment]
**Comment by misterx on 19 Feb 2016 10:26 UTC**

Nevermind, I got the answer.

**Comment by aircrack-ng** [comment]
**Comment by anonymous on 19 Feb 2016 10:26 UTC**

There's also a github issue.

https://github.com/aircrack-ng/aircrack-ng/issues/79

**Comment by aircrack-ng** [comment]
**Modified by misterx on 19 Feb 2016 10:26 UTC**

**Comment by aircrack-ng** [comment]
**Comment by misterx on 19 Feb 2016 10:26 UTC**

Might be linked to #1701.

**Comment by aircrack-ng** [comment]
**Comment by samueloph on 19 Feb 2016 10:26 UTC**

I've just confirmed this is linked to #1701, as disabling fortify fixes the problem, that would explain why this only showed up on aircrack-ng's packaging, the majority of people don't bother with hardening flags.

I think we should focus the discussion on this ticket (as its the oldest).

For the packaging, we could workaround by disabling fortify for the whole aircrack-ng, which doesn't looks good, but then if we don't do that, we end up with a malfunctioning aircrack-ng binary (doesn't pass the Harkonen check).

**Comment by aircrack-ng** [comment]
**Comment by misterx on 19 Feb 2016 10:26 UTC**

Does https://patch-diff.githubusercontent.com/raw/aircrack-ng/aircrack-ng/pull/57.diff fix the issue?

**Comment by aircrack-ng** [comment]
**Comment by samueloph on 19 Feb 2016 10:26 UTC**

Yes, it does :)

Patched: Passes at least 20 consecutive Harkonen tests.
Non-Patched: Segfault randomly, from 10 checks at least 1 segfaults.

**Comment by aircrack-ng** [comment]
**Comment by misterx on 19 Feb 2016 10:26 UTC**

Fixed in aircrack-ng/aircrack-ng@e63ce86096b980c34f2d85a43bb9c73171213bc1.
